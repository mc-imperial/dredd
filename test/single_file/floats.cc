int main() {
  double a = 1.0;
  double x = 5.32;
  x += 0.5;
  float y = 64343.7;
  y -= 1.2;
  double z = x * 5.5;
  return z + x;
}
