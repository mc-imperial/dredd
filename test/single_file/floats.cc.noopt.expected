#include <cinttypes>
#include <cstddef>
#include <functional>
#include <string>

static bool __dredd_some_mutation_enabled = true;
static bool __dredd_enabled_mutation(int local_mutation_id) {
  static bool initialized = false;
  static uint64_t enabled_bitset[2];
  if (!initialized) {
    bool some_mutation_enabled = false;
    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable != nullptr) {
      std::string contents(dredd_environment_variable);
      while (true) {
        size_t pos = contents.find(",");
        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
        if (!token.empty()) {
          int value = std::stoi(token);
          int local_value = value - 0;
          if (local_value >= 0 && local_value < 81) {
            enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
            some_mutation_enabled = true;
          }
        }
        if (pos == std::string::npos) {
          break;
        }
        contents.erase(0, pos + 1);
      }
    }
    initialized = true;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return (enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64))) != 0;
}

static int __dredd_replace_expr_int(std::function<int()> arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 0;
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return 1;
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -1;
  return arg();
}

static float& __dredd_replace_binary_operator_SubAssign_float_double(std::function<float&()> arg1, std::function<double()> arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1() -= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1() += arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1() = arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1() /= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1() *= arg2();
  return arg1() -= arg2();
}

static float __dredd_replace_expr_float(std::function<float()> arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return -(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return 0.0;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 1.0;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return -1.0;
  return arg();
}

static double& __dredd_replace_binary_operator_AddAssign_double_double(std::function<double&()> arg1, std::function<double()> arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1() += arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1() = arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1() /= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1() *= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1() -= arg2();
  return arg1() += arg2();
}

static double __dredd_replace_expr_double_lvalue(std::function<double&()> arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --(arg());
  return arg();
}

static double __dredd_replace_expr_double(std::function<double()> arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return -(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return 0.0;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return 1.0;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return -1.0;
  return arg();
}

static double __dredd_replace_binary_operator_Mul_double_double(std::function<double()> arg1, std::function<double()> arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1() * arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1() + arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1() / arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1() - arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return arg2();
  return arg1() * arg2();
}

static double __dredd_replace_binary_operator_Add_double_double(std::function<double()> arg1, std::function<double()> arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1() + arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1() / arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1() * arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1() - arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return arg2();
  return arg1() + arg2();
}

int main() {
  double a = __dredd_replace_expr_double([&]() -> double { return 1.0; }, 0);
  double x = __dredd_replace_expr_double([&]() -> double { return 5.32; }, 4);
  if (!__dredd_enabled_mutation(16)) { __dredd_replace_binary_operator_AddAssign_double_double([&]() -> double& { return static_cast<double&>(x); } , [&]() -> double { return static_cast<double>(__dredd_replace_expr_double([&]() -> double { return 0.5; }, 8)); }, 12); }
  float y = __dredd_replace_expr_float([&]() -> float { return __dredd_replace_expr_double([&]() -> double { return 64343.7; }, 17); }, 21);
  if (!__dredd_enabled_mutation(33)) { __dredd_replace_binary_operator_SubAssign_float_double([&]() -> float& { return static_cast<float&>(y); } , [&]() -> double { return static_cast<double>(__dredd_replace_expr_double([&]() -> double { return 1.2; }, 25)); }, 29); }
  double z = __dredd_replace_expr_double([&]() -> double { return static_cast<double>(__dredd_replace_binary_operator_Mul_double_double([&]() -> double { return static_cast<double>(__dredd_replace_expr_double([&]() -> double { return static_cast<double>(__dredd_replace_expr_double_lvalue([&]() -> double& { return static_cast<double&>(x); }, 34)); }, 36)); } , [&]() -> double { return static_cast<double>(__dredd_replace_expr_double([&]() -> double { return 5.5; }, 40)); }, 44)); }, 49);
  if (!__dredd_enabled_mutation(80)) { return __dredd_replace_expr_int([&]() -> int { return static_cast<int>(__dredd_replace_expr_double([&]() -> double { return static_cast<double>(__dredd_replace_binary_operator_Add_double_double([&]() -> double { return static_cast<double>(__dredd_replace_expr_double([&]() -> double { return static_cast<double>(__dredd_replace_expr_double_lvalue([&]() -> double& { return static_cast<double&>(z); }, 53)); }, 55)); } , [&]() -> double { return static_cast<double>(__dredd_replace_expr_double([&]() -> double { return static_cast<double>(__dredd_replace_expr_double_lvalue([&]() -> double& { return static_cast<double&>(x); }, 59)); }, 61)); }, 65)); }, 70)); }, 74); }
}
